_posts/2024-01-25-azure-without-secrets-part-2.md

```diff
@@ -0,0 +1,380 @@
+        }
+      ]
+      // ...
+    }
+    // ...
+  }
+}
+```
+
+#### Add NuGet packages
+We need two packages; `Azure.Storage.Blobs` and `Azure.Identity`. `Azure.Storage.Blobs` provides the Azure storage blob client implementations. `Azure.Identity` allows us to create a `DefaultAzureCredential` object, which inherits the `TokenCredential` class. Different Azure client classes accept a `TokenCredential` object as a parameter, which is then used as a token provider.
+
+`DefaultAzureCredential` checks different options, whether you've logged in to Azure CLI, Visual Studio or some other supported application, and then utilizes that identity. It also checks whether the host has managed identity enabled and that's what makes this work in Azure. A complete list of credentials it tries can be found [here](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet#definition). **Remember to log in to at least one of the listed applications for localhost development to work!**
+
+```shell
+dotnet add package Azure.Identity
+dotnet add package Azure.Storage.Blobs
+```
+
+#### Inject client and access blob storage
+The only thing left is to inject the client through dependency injection in `Program.cs` and then utilize it. Since we already have created a container (`data`) and we want to list and read blobs in that container, we're going to add `BlobContainerClient`  injection. `BlobContainerClient` has different constructors, but we'll use the one that accepts container `Uri` and `TokenCredential`.
+
+```csharp
+var builder = WebApplication.CreateBuilder(args);
+
+var containerAddress = string.Join('/', builder.Configuration["BlobEndpoint"].TrimEnd('/'), "data");
+builder.Services.AddSingleton(new BlobContainerClient(new Uri(containerAddress), new DefaultAzureCredential()));
+
+var app = builder.Build();
+```
+
+In .NET minimal API we can access the injected `BlobContainerClient` by adding it to handler parameters with `FromServices` attribute.
+```csharp
+app.MapGet("/", async ([FromServices] BlobContainerClient containerClient) =>
+{
+    // Do something with the containerClient here
+});
+```
+
+### Disable key-based auth (optional)
+You can disable key-based access to a storage account by setting `allowSharedKeyAccess` *false* in the storage account's `properties` in Bicep.
+
+```bicep
+resource storageAccount 'Microsoft.Storage/storageAccounts@2022-09-01' = {
+  name: storageAccountName
+  location: location
+  properties: {
+    allowSharedKeyAccess: false
+    // ...
+  }
+}
+```
+
+## App Service to Azure SQL Database
+In this example I'll show you how to use identity-based authentication to access data in Azure SQL Database. Granting access to SQL databases is a bit different compared to other examples. We can't use RBAC, because there are no data plane roles for SQL databases. Instead, we need to create a database user for the managed identity and grant database access to that user.
```

Comment:
Tähänki ehkä sama, että "in the following..." tms

Reply:
Jätin nää esimerkkien alut silleen niinku ne oli, kun tosiaan on kolme erillistä esimerkkiä.